ament_cpplint/ament_cpplint/main.py

```diff
@@ -199,27 +205,57 @@ def custom_error(filename, linenum, category, confidence, message):
     return 1 if _cpplint_state.error_count else 0
 
 
-def get_file_groups(paths, extensions):
+def is_file_or_directory_to_exclude(path_to_file, excludes):
+    if not os.path.isdir(path_to_file):
+        path, file_name = os.path.split(path_to_file)
+    else:
+        path = path_to_file
+        file_name = ''
+    path_to_check = pathlib.Path(path)
+
+    def check_path_parents(parents, exclude):
+        for p in parents:
+            if p.match(exclude):
+                return True
+        return False
+
+    for excl in excludes:
+        if (
+            excl == file_name or
+            path_to_check.match(excl) or
+            check_path_parents(path_to_check.parents, excl)
+        ):
+            return True
+    return False
+
+
+def get_file_groups(paths, extensions, excludes):
     # dict mapping root path to files
     groups = {}
     for path in paths:
         if os.path.isdir(path):
             for dirpath, dirnames, filenames in os.walk(path):
+                if is_file_or_directory_to_exclude(dirpath, excludes):
+                    continue
+
                 if 'AMENT_IGNORE' in dirnames + filenames:
                     dirnames[:] = []
                     continue
                 # ignore folder starting with . or _
                 dirnames[:] = [d for d in dirnames if d[0] not in ['.', '_']]
+                dirnames[:] = [d for d in dirnames if d not in excludes]
```

Comment:
`dirnames` only contains the names of the subdirectories of `dirpath`. As such the condition `not in excludes` is incorrect again - for the same reasons as the previous comments.